// resolve-dependencies/lib/loader-pool.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const file_1 = require("./file");
const path_1 = require("path");
const worker_1 = require("./worker");
const node_builtins_1 = __importDefault(require("./node-builtins"));
const os_1 = require("os");
class Loader {
    constructor(options) {
        var _a;
        this.options = options;
        this.pool = [];
        this.ended = false;
        this.currentWorker = 0;
        const cores = ((_a = os_1.cpus()) !== null && _a !== void 0 ? _a : []).length;
        this.size = Number(process.env.RESOLVE_DEPENDENCIES_CPUS) || (cores > 2 ? cores - 1 : 1);
        this.starting = [...Array(this.size)].map((x) => new worker_1.WorkerThread({ taskConccurency: 100 }));
        this.workerOptions = { ...options, files: {} };
    }
    setup() {
        if (this.initializing) {
            return this.initializing;
        }
        this.ended = false;
        //initailize all, but only wait for the first one
        return (this.initializing = Promise.race(this.starting.map((x) => {
            return x
                .sendMessage({
                modulePath: require.resolve('./node-loader'),
                contextName: 'node-loader',
                options: this.workerOptions,
            })
                .then(() => {
                if (this.ended) {
                    x.end();
                }
                this.pool.push(x);
            });
        })).then(() => undefined));
    }
    quit() {
        this.ended = true;
        this.pool.forEach((x) => x.end());
    }
    getWorker() {
        //todo wait for workers to warm up
        const worker = this.pool[this.currentWorker++];
        if (this.currentWorker === this.pool.length) {
            this.currentWorker = 0;
        }
        return worker;
    }
    loadEntry(workingDirectory, request, files = {}, warnings = []) {
        const mainFile = file_1.ensureDottedRelative(workingDirectory, path_1.resolve(workingDirectory, request));
        return this.load(workingDirectory, mainFile, files, warnings).then((entry) => {
            return { entry, files, warnings };
        }, (e) => {
            throw e;
        });
    }
    async load(cd, request, files = {}, warnings, context) {
        const worker = this.getWorker();
        const options = {};
        if (context) {
            Object.assign(options, this.workerOptions, { context });
        }
        else {
            Object.assign(options, this.workerOptions);
        }
        const file = await worker.sendMessage({
            contextName: 'node-loader',
            method: 'load',
            args: [cd, request, options],
        });
        if ('warning' in file) {
            warnings.push(file.warning);
            return null;
        }
        if (files[file.absPath] !== undefined) {
            return files[file.absPath];
        }
        else {
            files[file.absPath] = file;
        }
        let packageGlobs = undefined;
        if (file.moduleRoot && file.package && file.package.files) {
            packageGlobs = file.package.files;
        }
        if (!packageGlobs && file.belongsTo && file.belongsTo.package && file.belongsTo.package.files) {
            packageGlobs = file.belongsTo.package.files;
        }
        const fileDir = path_1.dirname(file.absPath);
        const ctx = {
            moduleRoot: file.moduleRoot || (file.belongsTo && file.belongsTo.moduleRoot) || undefined,
            package: file.package,
            expanded: Boolean(file.contextExpanded),
            globs: packageGlobs,
        };
        await Promise.all(Object.keys(file.deps).map((req) => {
            if (~node_builtins_1.default.indexOf(req)) {
                return (file.deps[req] = null);
            }
            return this.load(fileDir, req, files, warnings, ctx).then((dep) => {
                file.deps[req] = dep;
                if ((file.moduleRoot || file.belongsTo) && dep && !dep.moduleRoot) {
                    const owner = file.moduleRoot ? file : file.belongsTo;
                    dep.belongsTo = owner;
                }
            });
        }));
        return file;
    }
}
exports.Loader = Loader;
