// resolve-dependencies/lib/node-loader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = require("fs");
const globby_1 = __importDefault(require("globby"));
const path_1 = require("path");
const gather_deps_1 = require("./gather-deps");
const file_1 = require("./file");
const enhanced_resolve_1 = require("enhanced-resolve");
const readFile = fs_1.promises.readFile;
const fileSystem = new enhanced_resolve_1.CachedInputFileSystem(new enhanced_resolve_1.NodeJsInputFileSystem(), 4000), resolver = enhanced_resolve_1.ResolverFactory.createResolver({
    extensions: ['.js', '.json', '.node'],
    fileSystem: new enhanced_resolve_1.CachedInputFileSystem(new enhanced_resolve_1.NodeJsInputFileSystem(), 4000),
}), syncResolver = enhanced_resolve_1.ResolverFactory.createResolver({
    extensions: ['.js', '.json', '.node'],
    useSyncFileSystemCalls: true,
    fileSystem,
}), moduleGlob = ['**/*', '!node_modules', '!test'], defaultOptions = { loadContent: true, expand: 'none', isEntry: false };
function resolveSync(from, request) {
    const result = {
        absPath: '',
        pkgPath: '',
        pkg: null,
        warning: '',
    };
    syncResolver.resolve({}, from, request, {}, (err, path, data) => {
        if (err) {
            result.warning = err.message;
            return;
        }
        result.absPath = path;
        result.pkgPath = data.descriptionFilePath;
        result.pkg = data.descriptionFileData;
        return;
    });
    return result;
}
exports.resolveSync = resolveSync;
function resolve(from, request) {
    const result = {
        absPath: '',
        pkgPath: '',
        pkg: null,
        warning: '',
    };
    return new Promise((resolve, reject) => {
        resolver.resolve({}, from, request, {}, (err, path, data) => {
            if (err) {
                result.warning = err.message;
                return resolve(result);
            }
            result.absPath = path;
            result.pkgPath = data.descriptionFilePath;
            result.pkg = data.descriptionFileData;
            resolve(result);
        });
    });
}
exports.resolve = resolve;
async function expand(file, fileDir, baseDir, globs) {
    const files = await globby_1.default(globs || moduleGlob, { cwd: baseDir });
    files
        .map((dep) => file_1.ensureDottedRelative(fileDir, path_1.join(baseDir, dep)))
        .filter((relDep) => file.absPath !== path_1.join(baseDir, relDep))
        .forEach((relDep) => {
        file.deps[relDep] = file.deps[relDep] || null;
    });
    const currentDeps = Object.keys(file.deps);
    file.package &&
        file.package.dependencies &&
        Object.keys(file.package.dependencies || {}).forEach((dependency) => {
            if (!currentDeps.some((curDep) => curDep.startsWith(dependency))) {
                file.deps[dependency] = file.deps[dependency] || null;
            }
        });
}
async function load(workingDirectory, request, options = defaultOptions) {
    const { absPath, pkg, pkgPath, warning } = await resolve(workingDirectory, request);
    if (!absPath) {
        return { warning: warning };
    }
    const file = file_1.createFile(absPath), isJs = absPath.endsWith('.js') || absPath.endsWith('.mjs') || options.isEntry;
    file.absPath = absPath;
    if (isJs || absPath.endsWith('json')) {
        file.contents = await readFile(absPath, 'utf-8');
    }
    if (isJs) {
        try {
            const parseResult = gather_deps_1.gatherDependencies(file.contents, absPath.endsWith('.mjs') || undefined);
            Object.assign(file.deps, parseResult.deps);
            file.variableImports = parseResult.variable;
        }
        catch (e) {
            return { warning: `Error parsing file: "${file.absPath}"\n${e.stack}` };
        }
    }
    const fileDir = path_1.dirname(file.absPath);
    if (file_1.isNodeModule(request) && pkg && pkgPath) {
        const pkgDir = (file.moduleRoot = path_1.dirname(pkgPath));
        file.package = pkg;
        file.deps[file_1.ensureDottedRelative(fileDir, pkgPath)] = null;
        if (options.expand === 'all' || (options.expand === 'variable' && file.variableImports)) {
            await expand(file, fileDir, pkgDir, file.package.files);
            file.contextExpanded = true;
        }
    }
    else if (options.expand === 'variable' &&
        file.variableImports &&
        options.context &&
        options.context.moduleRoot &&
        !options.context.expanded) {
        await expand(file, fileDir, options.context.moduleRoot, options.context.globs);
        file.contextExpanded = true;
    }
    if (!options.loadContent) {
        file.contents = null;
    }
    return file;
}
exports.load = load;
